Update email upsell button to link directly to email addition page

wongasy opened the pull request:
#### Proposed Changes

Currently, the **"Create a new mailbox"** button in the Embedded Inbox links out to **Upgrades > Email** `/email/:site`, where users will need to make two more clicks (select a domain, click on "Add new mailboxes") before they get to the email addition screen:

<img width="1122" alt="Screen Shot 2022-06-28 at 3 20 00 PM" src="https://user-images.githubusercontent.com/104910361/176271403-9cbaeb41-73bc-4806-b9ff-330ee30220a9.png">

The proposed change improves UX by linking users **directly to the email addition page**, provided that there is **a single domain** associated with the domain.

<img width="1102" alt="Screen Shot 2022-06-28 at 3 20 18 PM" src="https://user-images.githubusercontent.com/104910361/176268738-fc3498cc-c95c-4cb0-9040-8defcce6a10d.png">

Also took the opportunity to separate out `NewMailboxUpsell` into its own file, so things look cleaner with the additional logic for generating a different click through URL for the button. 

Ideally, I would love to have `<NewMailboxUpsell>` added via `/inbox/index.jsx` rather than adding it as part of `<MailboxSelectionList>`, since it isn't really a mailbox list item and can be placed anywhere on the page. But for now, I left upsell notice as-is, since we'd want the notice to appear at the same time as the mailbox list.

#### Testing Instructions

**Scenario A:** User has multiple domains associated with the site
1. Have multiple domains
2. Subscribe to an email service
3. Click on **"Inbox"** in the sidebar
4. Click on **"Create a new mailbox"** button

**Expected outcome:** Users land on the **Upgrades > Email** page `/email/:site`, where they need to select a domain.

---

**Scenario B:** User has a single domain associated with the site, and is subscribed to Professional Email
1. Have a single domain
2. Subscribe to Professional Email service
3. Click on **"Inbox"** in the sidebar
4. Click on **"Create a new mailbox"** button

**Expected outcome:** Users land on the email addition page for Titan `/email/:domain/titan/new/:site?source=inbox`

---

**Scenario C:** User has a single domain associated with the site, and is subscribed to Google Workspace
1. Have a single domain
2. Subscribe to Google Workspace service
3. Click on **"Inbox"** in the sidebar
4. Click on **"Create a new mailbox"** button

**Expected outcome:** Users land on the email addition page for Google `/email/:domain/google-workspace/add-users/:site?source=inbox`

Commit:
Update email inbox upsell button to link directly to email addition page.

wongasy commented on client/my-sites/email/inbox/mailbox-selection-list/index.jsx:
```diff
@@ -131,14 +131,16 @@ MailboxItem.propType = {
 		recordInboxNewMailboxUpsellClickEvent();
-		page( emailManagement( selectedSiteSlug, null, null, { source: INBOX_SOURCE } ) );
+		page(
+			emailManagementEdit( selectedSiteSlug, domain, 'titan/new', null, { source: INBOX_SOURCE } )
```
Getting the `domain` passed from the [mailbox object](https://github.com/Automattic/wp-calypso/pull/65058/files#diff-826f692b78a2fbee078b7b3e0372cb4c986011d0f98d577351077a2a68d64438R280) instead of getting it from `selectedSite` because the free domain gets returned as `selectedSite.domain` in **Scenario C** (free domain is set as primary address despite having a custom domain in place), and we end up with an invalid URL `/email/:site/manage/:site`.

olaseni replied:
Note that this solution also only works for Professional Email, and not Google Workspace.

fredrikekelund commented on client/my-sites/email/inbox/mailbox-selection-list/index.jsx:
```diff
@@ -131,14 +131,16 @@ MailboxItem.propType = {
 	const handleCreateNewMailboxClick = useCallback( () => {
 		recordInboxNewMailboxUpsellClickEvent();
-		page( emailManagement( selectedSiteSlug, null, null, { source: INBOX_SOURCE } ) );
+		page(
```
I think it would be nice to take this opportunity to pass this path as an `href` prop to `Button` instead of using the `page` function to navigate programmatically.

We could still use the `onClick` handler for tracking the click event.

wongasy replied:
I left `page` in place. I don't actually know what `page` does 😆 but I wonder if it applies some sort of delay to ensure our tracking event gets sent before navigating away from the current page? Is `page` a third party library?

fredrikekelund replied:
I followed up in a new thread!

fredrikekelund commented on client/my-sites/email/inbox/mailbox-selection-list/index.jsx:
```diff
@@ -173,7 +175,7 @@ const MailboxItems = ( { mailboxes } ) => {
 			) ) }
 
-			<NewMailboxUpsell />
+			<NewMailboxUpsell domain={ domain } />
```
Seeing as we're using the `mailboxes` array to extract the domain, I think we could change the value here to `mailboxes[ 0 ]?.domain`. That way we don't have to pass the prop explicitly to this component.

fredrikekelund reviewed (changes requested):
This is a nice UX improvement! 👍 

I left a couple of comments. The most important thing is probably that we should consider a scenario where a user has multiple custom domains connected to a single site.

olaseni commented:
> The most important thing is probably that we should consider a scenario where a user has multiple custom domains connected to a single site.

Perhaps we could borrow the approach from #64933 and make a case exclusively for when the user has a single custom domain in that site?

Commit:
Decoupled NewMailboxUpsell from MailboxSelectionList, generate email upsell URL by email subscription for better UX

wongasy commented:
Thank you @olaseni and @fredrikekelund for the reviews!  Great catch on the use cases for multiple domains and Google Workspace! I ended up making a fair amount of changes to address those scenarios, and also did a bit of refactoring. Since there are quite a lot of changes (compared to my initial commit), I re-wrote the PR message as well :)

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.jsx:
```diff
@@ -0,0 +1,66 @@
+	const selectedSiteSlug = selectedSite?.slug;
+
+	let upsellURL = '';
+	if ( 1 === domains.length ) {
```
Kudos for going the classic WordPress route here and choosing a Yoda conditional 😉

I know that the [JS coding guidelines](https://wpcalypso.wordpress.com/devdocs/docs/coding-guidelines/javascript.md#yoda-conditions) are ambivalent on the matter, but my experience is that they are in fact quite rare in Calypso and for the back-end, there was a [recent change](https://make.wordpress.org/core/2022/06/14/upcoming-disallow-assignments-in-conditions-and-remove-the-yoda-condition-requirement-for-php/) in the guidelines where they are now actively discouraged (the article provides some nice background as well).

I would encourage you not to use them, but it's no requirement.

wongasy replied:
Thanks for keeping me updated on the guidelines :) Very helpful!

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.jsx:
```diff
@@ -0,0 +1,66 @@
+	let upsellURL = '';
+	if ( 1 === domains.length ) {
+		// User has one single domain, determine subscribed email provider.
+		if ( 'active' === domains[ 0 ]?.titanMailSubscription?.status ) {
```
We should use the `hasTitanMailWithUs` function here (located in `calypso/lib/titan`)

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.jsx:
```diff
@@ -0,0 +1,66 @@
+			upsellURL = emailManagementEdit( selectedSiteSlug, domains[ 0 ]?.domain, 'titan/new', null, {
+				source: INBOX_SOURCE,
+			} );
+		} else if ( 'active' === domains[ 0 ]?.googleAppsSubscription?.status ) {
```
We should use the `hasGSuiteWithUs` function here (located in `calypso/lib/gsuite`)

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.jsx:
```diff
@@ -0,0 +1,66 @@
+	if ( 1 === domains.length ) {
+		// User has one single domain, determine subscribed email provider.
+		if ( 'active' === domains[ 0 ]?.titanMailSubscription?.status ) {
+			upsellURL = emailManagementEdit( selectedSiteSlug, domains[ 0 ]?.domain, 'titan/new', null, {
```
Seeing as `MailboxSelectionList` is [only rendered if there is at least one domain](https://github.com/Automattic/wp-calypso/blob/af50f9d5b36fb6e0b63fa7eae56afdbc099dff26/client/my-sites/email/inbox/index.jsx#L121), I don't think we actually need to use optional chaining after `domains[ 0 ]`, right?

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.jsx:
```diff
@@ -0,0 +1,66 @@
+	}
+
+	// Upsell URL not determined because user has multiple domains or none of the domains has email service subscription.
+	if ( '' === upsellURL ) {
```
I think it would improve readability to either make this an `else` condition of the `1 === domains.length` conditional, or alternatively to make this assignment to the `upsellURL` variable the default one.

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.jsx:
```diff
@@ -0,0 +1,66 @@
+		upsellURL = emailManagement( selectedSiteSlug, null, null, { source: INBOX_SOURCE } );
+	}
+
+	const handleCreateNewMailboxClick = useCallback( () => {
```
I know this `useCallback` call was here previously, but I actually think we could take this opportunity to drop it. There are no performance gains from using it here and it increases complexity.

wongasy replied:
Agreed! Especially when we're dropping `page()` and adding the `href` to the button itself.

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.jsx:
```diff
@@ -0,0 +1,66 @@
+			<div className="new-mailbox-upsell">
+				<div className="new-mailbox-upsell__messages">
+					<h2>{ translate( 'Need another mailbox?' ) }</h2>
+					<div>{ translate( 'Create new and activate immediately' ) }</div>
```
I know this also is identical to what we had previously, but this string just sounds a bit wonky to me. Should we change it to something like `Create a new one and activate it immediately`.

@stephanethomas @AllTerrainDeveloper @olaseni WDYT?

wongasy replied:
I agree that the suggested copy sounds better!

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.jsx:
```diff
@@ -0,0 +1,66 @@
+					<div>{ translate( 'Create new and activate immediately' ) }</div>
+				</div>
+				<div className="new-mailbox-upsell__cta">
+					<Button onClick={ handleCreateNewMailboxClick }>
```
I still think it would be nice to use an `href` prop here 😊

`page` is this tiny client-side routing library that we use in Calypso. The main function can either define routes (call it with one string and one callback) or navigate to routes (call it with just a string). Here's the [API reference](https://www.npmjs.com/package/page#api).

If we preserve the `onClick` handler (but remove the `page` call in that function) and add an `href` prop – I can confirm that the same events are tracked in Calypso.

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/style.scss:
```diff
@@ -0,0 +1,72 @@
+@import '@wordpress/base-styles/breakpoints';
+@import '@wordpress/base-styles/mixins';
+
+.new-mailbox-upsell__container {
```
Nice adaptation of the class names 👍

fredrikekelund reviewed (changes requested):
Making very good progress here! 👍

I left a number of comments, but most of them should hopefully be quite straightforward to address.

Moving `NewMailboxUpsell` into its own file also seems like a reasonable decision now that it's grown a bit.

Commit:
Simplified email subscription check using helper functions and made tweaks to improve code readability

Commit:
Adjust back/cancel button links based on context of email addition user journey to improve UX

wongasy commented:
@fredrikekelund I have addressed all your feedbacks :) Thank you again for the review, all suggestions were good!

I also made another UX update so that the "Back" and "Cancel" button in the email addition screen aligns with the context of how user got there. If the `source` is `inbox`, then `goToEmail()` should link users back to the **Inbox** page as opposed to the **Email Management** page, given the **Inbox** is actually where the user came from.

fredrikekelund commented on client/my-sites/email/gsuite-add-users/index.jsx:
```diff
@@ -174,13 +174,15 @@ class GSuiteAddUsers extends Component {
+			url = emailManagementInbox( selectedSite.slug );
+		}
+
+		page( url );
```
This is a good initiative, but with the very recent completion of the "unify mailboxes" project, this component is no longer in use.

You should rebase this branch and move this implementation to the `goToEmail` function in `client/my-sites/email/add-mailboxes/index.tsx`.

fredrikekelund commented on client/my-sites/email/titan-add-mailboxes/index.jsx:
```diff
@@ -82,14 +86,18 @@ class TitanAddMailboxes extends Component {
+			url = emailManagementInbox( this.props.selectedSite.slug );
+		}
+
+		page( url );
```
Same here as for the `GSuiteAddUsers` component 🙂

fredrikekelund commented:
@wongasy good idea with the modified logic around the cancel/back operation! I left some comments about the implementation, though.

Hopefully we should be good to merge after you've addressed those 👍

olaseni commented on client/my-sites/email/inbox/new-mailbox-upsell/style.scss:
```diff
@@ -0,0 +1,72 @@
+		}
+
+		@include break-large {
+			width: 660px;
```
Why do we need this hard width?

wongasy replied:
Without this, the upsell CTA doesn't line up properly with the mailbox list: https://github.com/Automattic/wp-calypso/blob/8ca55b87d2292b4ece8d2b2549f7103ff5c1afa2/client/my-sites/email/inbox/mailbox-selection-list/style.scss#L123

olaseni commented on client/my-sites/email/inbox/new-mailbox-upsell/index.jsx:
```diff
@@ -0,0 +1,57 @@
+	let upsellURL = emailManagement( selectedSiteSlug, null, null, { source: INBOX_SOURCE } );
+
+	// User has one single domain, determine subscribed email provider.
+	if ( domains.length === 1 ) {
+		if ( hasTitanMailWithUs( domains[ 0 ] ) ) {
+			upsellURL = emailManagementEdit( selectedSiteSlug, domains[ 0 ].domain, 'titan/new', null, {
+				source: INBOX_SOURCE,
+			} );
+		} else if ( hasGSuiteWithUs( domains[ 0 ] ) ) {
+			upsellURL = emailManagementEdit(
+				selectedSiteSlug,
+				domains[ 0 ].domain,
+				'google-workspace/add-users',
+				null,
+				{ source: INBOX_SOURCE }
+			);
+		}
+	}
```
```suggestion
	const getSectionSlug = ( domain ) => {
		if ( hasTitanMailWithUs( domain ) ) {
			return 'titan/new';
		}

		if ( hasGSuiteWithUs( domain ) ) {
			return 'google-workspace/add-users';
		}	
		
		return null
	};

	if ( domains.length === 1 && getSectionSlug( domain[ 0 ] ) ) {
		upsellURL = emailManagementEdit( selectedSiteSlug, domains[ 0 ].domain, getSectionSlug( domain [ 0 ] ), null, {
				source: INBOX_SOURCE,
			} );
	}
```

Nit: would this be more readable? Less logic duplication?

wongasy replied:
I like the idea of calculating a `slug` and removed the duplicated `emailManagementEdit()` :) I've adapted a similar approach, but without introducing a new helper function `getSectionSlug()`. I'd like to avoid having to call it twice, when we could have just stored the calculated value in a variable.

olaseni reviewed (changes requested):
This needs to be rebased like @fredrikekelund mentioned as referenced files no longer exist, then we should probably test this again.

olaseni commented on client/my-sites/email/inbox/new-mailbox-upsell/index.jsx:
```diff
@@ -0,0 +1,57 @@
+import { Button } from '@automattic/components';
```
I forgot to mention - we now always add TS/TSX files rather than JSX.

Commit:
Adjusted upsell button url logic to improve readability

Commit:
Renamed uNewMailboxUpsell component .jsx file to .tsx

wongasy commented:
For some reason, the build is failing at `Unit tests`. I'm not seeing any explicit message that explains what went wrong... I've tried rebuilding a few times without success. Wonder if this was because I renamed the `email/inbox/new-mailbox-upsell/index.jsx` to `.tsx`?

fredrikekelund commented:
> For some reason, the build is failing at Unit tests. I'm not seeing any explicit message that explains what went wrong...

@wongasy I also had some trouble deciphering the TeamCity logs to begin with, but this is a type error in `client/my-sites/email/inbox/new-mailbox-upsell/index.tsx` that you can find some more details about under the Build Log tab ([ref](https://teamcity.a8c.com/viewLog.html?buildId=8164173&buildTypeId=calypso_RunAllUnitTests&tab=buildLog&_focus=480)).

Commit:
Bind property tdomains to proper data type

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.tsx:
```diff
@@ -0,0 +1,60 @@
+		if ( hasTitanMailWithUs( domains[ 0 ] ) ) {
+			slug = 'titan/new';
+		} else if ( hasGSuiteWithUs( domains[ 0 ] ) ) {
+			upsellURL = 'google-workspace/add-users';
```
This should be `slug = 'google-workspace/add-users';`, right?

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.tsx:
```diff
@@ -0,0 +1,60 @@
+	if ( domains.length === 1 ) {
+		const domainItem = domains[ 0 ];
+
+		let slug = null;
```
Nit, but if we make the default value an empty string instead of `null`, Typescript will automatically assign the type `string` to the variable, which should be appropriate.

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.tsx:
```diff
@@ -0,0 +1,60 @@
+ */
+import './style.scss';
+
+const NewMailboxUpsell = ( { domains }: { domains: Array< ResponseDomain > } ) => {
```
Nit, but the notation `ResponseDomain[]` is probably more common compared to `Array< ResponseDomain >` (at least in Calypso). So long as we don't have any specific reason for using `Array<>` (e.g. we are defining multiple allowed types like `Array< string | null >`), I think it makes sense to go with `ResponseDomain[]`.

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.tsx:
```diff
@@ -0,0 +1,60 @@
+	// By default, upsell CTA links to the email management landing page.
+	let upsellURL = emailManagement( selectedSiteSlug, null, null, { source: INBOX_SOURCE } );
+
+	// User has one single domain, determine \ email addition page URL based on subscribed email provider.
```
Another nit, but why the backslash in this comment? 🤔

fredrikekelund reviewed (changes requested):
Back button from the "add mailbox" page works well, `href` is in place on the button 👍 It looks good, but I found a few nitpicks and one blocker (variable name) before we can merge.

Commit:
Addressed revewer feedbacks

fredrikekelund commented on client/my-sites/email/inbox/new-mailbox-upsell/index.tsx:
```diff
@@ -0,0 +1,60 @@
+		const domainItem = domains[ 0 ];
+
+		let slug = '';
+		if ( hasTitanMailWithUs( domains[ 0 ] ) ) {
```
I missed this one previously, but since we have the `domainItem` variable here, we should probably take the opportunity to change this to `hasTitanMailWithUs( domainItem )` (same thing for `hasGSuiteWithUs`)

Commit:
Use defined variable for better readability

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7419598

Thank you @wongasy for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7419598).
